all build output should be captured

As far as I can tell, not all of the output from a build is captured by opam. I believe the issue is that a build section can contain multiple commands, but opam saves the output of only the last command. This makes debugging difficult.

As a concrete example, when installing from a test opam package, opam says:

```
biocaml.0.3.1 Downloading https://github.com/biocaml/biocaml/archive/0.3.1.tar.gz
Building biocaml.0.3.1:
  omake configure PREFIX=/Users/ashish/.opam/4.01.0 BUILD_TESTS=false BUILD_APP=false
  omake -j2
  omake install
[ERROR] The compilation of biocaml.0.3.1 failed.
Removing biocaml.0.3.1.
  ocamlfind remove biocaml
```

I reproduce the exact same commands as above:

```
$ wget --no-check-certificate https://github.com/biocaml/biocaml/archive/0.3.1.tar.gz
...

$ tar xzvf 0.3.1.tar.gz
...

$ omake configure PREFIX=/Users/ashish/.opam/4.01.0 BUILD_TESTS=false BUILD_APP=false
*** omake: reading OMakefiles
--- Checking for ocamlfind... (found /Users/ashish/.opam/4.01.0/bin/ocamlfind)
--- Checking for ocamlc.opt... (found /Users/ashish/.opam/4.01.0/bin/ocamlc.opt)
--- Checking for ocamlopt.opt... (found /Users/ashish/.opam/4.01.0/bin/ocamlopt.opt)
--- Checking whether ocamlc understands the "z" warnings... (yes)
conf: Building biocaml application
conf: Building biocaml tests
conf: Building biocaml benchmarks
*** omake: finished reading OMakefiles (0.05 sec)
- build . <configure>
+ global.configurable("BUILD_APP", "COMPILE_BIOCAML_APP")
User defined BUILD_APP: false
- build . <configure>
+ global.configurable("BUILD_TESTS", "COMPILE_BIOCAML_TESTS")
User defined BUILD_TESTS: false
- build . <configure>
+ global.configurable("PREFIX", "INSTALL_PREFIX")
User defined PREFIX: /Users/ashish/.opam/4.01.0
*** omake: done (0.06 sec, 0/0 scans, 1/1 rules, 10/226 digests)
```

I would like to see the above output when installing via opam. However, I can't find this output in any file under `~/.opam/4.01.0/build/biocaml.0.3.1`. Is there somewhere else I should look also? The output I do find is that of `omake -j2`, which is past the above configure step I'm trying to debug.

--- a/src/core/opamProcess.ml
+++ b/src/core/opamProcess.ml
@@ -27,7 +27,7 @@ type t = {
   p_metadata: (string * string) list;
 }
 
-let open_flags =  [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC; Unix.O_EXCL]
+let open_flags =  [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_APPEND]
 
 let output_lines oc lines =
   List.iter (fun line ->
@@ -81,7 +81,7 @@ let create ?info_file ?env_file ?stdout_file ?stderr_file ?env ?(metadata=[])
     let close_fd () = Unix.close fd in
     if verbose then (
       flush stderr;
-      let chan = Unix.open_process_out ("tee " ^ Filename.quote f) in
+      let chan = Unix.open_process_out ("tee -a " ^ Filename.quote f) in
       let close () =
         match Unix.close_process_out chan with
         | _ -> close_fd () in

--- a/src/core/opamSystem.ml
+++ b/src/core/opamSystem.ml
@@ -272,17 +272,19 @@ let print_stats () =
     OpamGlobals.msg "%d external processes called:\n  %s\n%!"
       (List.length l) (String.concat "\n  " (List.map (String.concat " ") l))
 
-let run_process ?verbose ?(env=default_env) ?name ?metadata command =
+let log_file name = match name with
+  | None   -> temp_file "log"
+  | Some n -> temp_file ~dir:(Sys.getcwd ()) n
+
+let log_cleanup r =
+  if not !OpamGlobals.debug then OpamProcess.clean_files r
+
+let run_process ?verbose ?(env=default_env) ~name ?metadata command =
   let chrono = OpamGlobals.timer () in
   runs := command :: !runs;
   match command with
   | []          -> invalid_arg "run_process"
   | cmd :: args ->
-
-    (* Set-up the log files *)
-    let name = match name with
-      | None   -> temp_file "log"
-      | Some n -> temp_file ~dir:(Sys.getcwd ()) n in
 
     (* Check that the command doesn't contain whitespaces *)
     if None <> try Some (String.index cmd ' ') with Not_found -> None then
@@ -295,8 +297,6 @@ let run_process ?verbose ?(env=default_env) ?name ?metadata command =
         | Some b -> b in
 
       let r = OpamProcess.run ~env ~name ~verbose ?metadata cmd args in
-      if OpamProcess.is_success r && not !OpamGlobals.debug then
-        OpamProcess.clean_files r;
       let str = String.concat " " (cmd :: args) in
       log "[%s] (in %.3fs) %s" (Filename.basename name) (chrono ()) str;
       r
@@ -305,29 +305,32 @@ let run_process ?verbose ?(env=default_env) ?name ?metadata command =
       command_not_found cmd
 
 let command ?verbose ?env ?name ?metadata cmd =
-  let r = run_process ?verbose ?env ?name ?metadata cmd in
-  if not (OpamProcess.is_success r) then
-    process_error r
+  let name = log_file name in
+  let r = run_process ?verbose ?env ~name ?metadata cmd in
+  if OpamProcess.is_success r then log_cleanup r
+  else process_error r
 
 let commands ?verbose ?env ?name ?metadata ?(keep_going=false) commands =
-  let err = ref None in
-  let command =
-    if keep_going then
-      (fun c ->
-        try command ?verbose ?env ?name ?metadata c with
-        | Process_error r -> if !err = None then err := Some r)
-    else
-      fun c -> command ?verbose ?env ?name ?metadata c
+  let name = log_file name in
+  let run = run_process ?verbose ?env ~name ?metadata in
+  let command r0 c =
+    if keep_going || OpamProcess.is_success r0 then
+      let r = run c in
+      (if OpamProcess.is_failure r0 then r0 else r)
+    else r0
   in
-  List.iter command commands;
-  match !err with
-  | Some err -> raise (Process_error err)
-  | None -> ()
+  match commands with
+  | [] -> ()
+  | c1::c ->
+    let r = List.fold_left command (run c1) c in
+    if OpamProcess.is_success r then log_cleanup r
+    else process_error r
 
 let read_command_output ?verbose ?env ?metadata cmd =
-  let r = run_process ?verbose ?env ?metadata cmd in
+  let name = log_file None in
+  let r = run_process ?verbose ?env ~name ?metadata cmd in
   if OpamProcess.is_success r then
-    r.OpamProcess.r_stdout
+    (log_cleanup r; r.OpamProcess.r_stdout)
   else
     process_error r
 

